alea/ces_source.py

```diff
@@ -0,0 +1,234 @@
+        return h
+
+    def _check_histogram(self, h: Hist1d):
+        """Check if the histogram has expected binning."""
+        # We only take 1d histogram in the ces axes
+        if not isinstance(h, Hist1d):
+            raise ValueError("Only Hist1d object is supported")
+        if self.ces_space.ndim != 1:
+            raise ValueError("Only 1d analysis space is supported")
+        if np.min(h.histogram) < 0:
+            raise AssertionError(
+                f"There are bins for source {self.templatename} with negative entries."
+            )
+
+        # check if the histogram contains the analysis space.
+        histogram_max = np.max(h.bin_edges)
+        histogram_min = np.min(h.bin_edges)
+        if self.min_e < histogram_min or self.max_e > histogram_max:
+            raise ValueError(
+                f"The histogram edge ({histogram_min},{histogram_max}) \
+                does not contain the analysis space ({self.min_e},{self.max_e})"
+            )
+
+    def _create_transformation(
+        self, transformation_type: Literal["efficiency", "smearing", "bias"]
+    ):
+        if self.config.get(f"apply_{transformation_type}", True):
+            parameters_key = f"{transformation_type}_parameters"
+            model_key = f"{transformation_type}_model"
+
+            if model_key not in self.config:
+                raise ValueError(f"{transformation_type.capitalize()} model is not provided")
+
+            if parameters_key not in self.config:
+                raise ValueError(f"{transformation_type.capitalize()} parameters are not provided")
+            else:
+                parameter_list = self.config[parameters_key]
+                # to get the values we need to iterate over the list and use self.config.get
+                combined_parameter_dict = {k: self.config.get(k) for k in parameter_list}
+
+            # Also take the peak_energy parameter if it is a mono smearing model
+            if "mono" in self.config[model_key]:
+                combined_parameter_dict["peak_energy"] = self.config["peak_energy"]
+
+            return Transformation(
+                parameters=combined_parameter_dict,
+                action=transformation_type,
+                model=self.config[model_key],
+            )
+        return None
+
+    def _transform_histogram(self, h: Hist1d):
+        # Create transformations for efficiency, smearing, and bias
+        efficiency_transformation = self._create_transformation("efficiency")
+        smearing_transformation = self._create_transformation("smearing")
+        bias_transformation = self._create_transformation("bias")
+
+        # Apply the transformations to the histogram
+        if efficiency_transformation is not None:
+            h = efficiency_transformation.apply_transformation(h)
+        if smearing_transformation is not None:
+            h = smearing_transformation.apply_transformation(h)
+        if bias_transformation is not None:
```

Comment:
Actually I think it's better to add in the smeared energy. Because the bias should in principle be caused by some area-dependent effect (like merging SEs and lone hits). For example if a monoenergetic source produces two S2s, 1.1e4 and 1.2 4 separately, the biases would be slightly different for them even if the true energies are the same.

Reply:
In general the bias should be a function of true parameters (I agree if we had a measurement of "true quanta released" we could define our bias in terms of that, and the CES could be considered an approximation of that, but it is only an approximation)